tokio-util/src/sync/cancellation_token.rs

```diff
@@ -253,42 +266,55 @@ impl CancellationToken {
     where
         F: Future,
     {
-        pin_project! {
-            /// A Future that is resolved once the corresponding [`CancellationToken`]
-            /// is cancelled or a given Future gets resolved. It is biased towards the
-            /// Future completion.
-            #[must_use = "futures do nothing unless polled"]
-            struct RunUntilCancelledFuture<'a, F: Future> {
-                #[pin]
-                cancellation: WaitForCancellationFuture<'a>,
-                #[pin]
-                future: F,
-            }
-        }
-
-        impl<'a, F: Future> Future for RunUntilCancelledFuture<'a, F> {
-            type Output = Option<F::Output>;
-
-            fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
-                let this = self.project();
-                if let Poll::Ready(res) = this.future.poll(cx) {
-                    Poll::Ready(Some(res))
-                } else if this.cancellation.poll(cx).is_ready() {
-                    Poll::Ready(None)
-                } else {
-                    Poll::Pending
-                }
-            }
+        RunUntilCancelledFuture {
+            cancellation: self.cancelled(),
+            future: fut,
         }
+        .await
+    }
 
+    /// Runs a future to completion and returns its result wrapped inside of an `Option`
+    /// unless the `CancellationToken` is cancelled. In that case the function returns
+    /// `None` and the future gets dropped.
+    ///
+    /// The function takes self by value.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is only cancel safe if `fut` is cancel safe.
+    pub async fn run_until_cancelled_owned<F>(self, fut: F) -> Option<F::Output>
+    where
+        F: Future,
+    {
         RunUntilCancelledFuture {
-            cancellation: self.cancelled(),
+            cancellation: self.cancelled_owned(),
             future: fut,
         }
         .await
     }
 }
 
+// ===== impl RunUntilCancelledFuture =====
+
+impl<C, F: Future> Future for RunUntilCancelledFuture<C, F>
```

Comment:
```suggestion
impl<C, F> Future for RunUntilCancelledFuture<C, F>
```

The bound on `Future` for `F` is already specified in the where clause.